scripts/research/b2_01_volatility_transition_lib.py

```
@@ -0,0 +1,1033 @@
+    base_pred = np.full(len(y), np.nan, dtype=np.float64)
+    cand_pred = np.full(len(y), np.nan, dtype=np.float64)
+    eligible = np.zeros(len(y), dtype=bool)
+
+    def add_record(j: int) -> None:
+        if not (
+            0 <= j < len(y)
+            and math.isfinite(float(y[j]))
+            and int(level[j]) >= 0
+        ):
+            return
+        l = int(level[j])
```

Ruff (0.16.2)

[error] Ambiguous variable name: `l`

(E741)